assertk/src/commonMain/kotlin/assertk/failure.kt

```diff
@@ -78,12 +78,28 @@ internal interface Failure {
 @PublishedApi
 internal inline fun <F: Failure, T> F.run(f: F.() -> T): T {
     pushFailure()
+    var otherException: Throwable? = null
     try {
         return f()
+    } catch (e: Throwable) {
+        if (e.isOutOfMemory()) {
+            throw e
+        }
+        otherException = e
     } finally {
         popFailure()
-        invoke()
+        if (otherException != null) {
+            try {
+                invoke()
+            } catch (e: Throwable) {
+                otherException.addSuppressed(e)
+            }
+            throw otherException
+        } else {
+            invoke()
+        }
     }
+    throw UnsupportedOperationException("unreachable!")
```

Comment:
not in love with having to check oom twice and the fact that kotlin's static analysis doesn't see the end of this function as unreachable, but not sure there's a better way.